guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+    |               --------- this data with an anonymous lifetime `'_`...
+145 |           Box::pin(async move {
+    |  _____________________________^
+146 | |             let result = tokio::time::timeout(
+147 | |                 self.duration,
+148 | |                 self.inner_handler.call(request),
+...   |
+155 | |             }
+156 | |         })
+    | |_________^ ...is captured here, requiring it to live as long as `'static`
+```
+
+The issue is that we're capturing a `&mut self` and moving it into an async
+block. That means the lifetime of our future is tied to the lifetime of `&mut
+self`. This doesn't work for us since we might want to run our response futures
+on multiple threads to get better performance or produce multiple response
+futures and run them all in parallel. That isn't possible if a reference to the
+handler lives inside the futures.
+
+Instead we need convert the `&mut self` into an owned `self`. That is exactly
+what `Clone` does:
+
+```rust
+// this must be `Clone` for `Timeout<T>` to be `Clone`
+#[derive(Clone)]
+struct RequestHandler;
+
+impl Handler for RequestHandler {
+    // ...
+}
+
+#[derive(Clone)]
+struct Timeout<T> {
+    inner_handler: T,
+    duration: Duration,
+}
+
+impl<T> Handler for Timeout<T>
+where
+    T: Handler + Clone,
+{
+    type Future = Pin<Box<dyn Future<Output = Result<Response, Error>>>>;
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        // get an owned clone of `&mut self`
+        let mut this = self.clone();
+
+        Box::pin(async move {
+            let result = tokio::time::timeout(
+                this.duration,
+                this.inner_handler.call(request),
+            ).await;
+
+            match result {
+                Ok(Ok(response)) => Ok(response),
+                Ok(Err(error)) => Err(error),
+                Err(_timeout) => Err(Error::timeout()),
+            }
+        })
+    }
+}
+```
+
+Note that cloning is very cheap in this case since `RequestHandler` doesn't have
+any data and `Timeout<T>` only adds a `Duration`.
+
+One step closer. We now get a different error:
+
+```
+error[E0310]: the parameter type `T` may not live long enough
+   --> src/lib.rs:149:9
+    |
+140 |   impl<T> Handler for Timeout<T>
+    |        - help: consider adding an explicit lifetime bound...: `T: 'static`
+...
+149 | /         Box::pin(async move {
+150 | |             let result = tokio::time::timeout(
+151 | |                 this.duration,
+152 | |                 this.inner_handler.call(request),
+...   |
+159 | |             }
+160 | |         })
+    | |__________^ ...so that the type `impl Future` will meet its required lifetime bounds
+```
+
+The problem now is that `T` can be any type whatsoever. It can even be a type
+that contains references like `Vec<&'a str>`. However that wont work for the
+same reason as before. We need the response future to have a `'static` lifetime
```

Comment:
This is because again our associated type is `Pin<Box<dyn Future + 'static>>` (if you don't include a lifetime its auto a `'static`. And again the same problem as above.